ledger/error/generator/lib/src/main/scala/com/daml/error/generator/ErrorCodeDocumentationGenerator.scala

```diff
@@ -8,239 +8,229 @@ import java.lang.reflect.Modifier
+  ): Seq[ErrorGroupDocItem] = {
+    val errorGroups = findInstancesOf[ErrorGroup](searchPackagePrefixes)
+    errorGroups.view.map(_.errorClass).groupBy(identity).collect {
       case (group, occurrences) if occurrences.size > 1 =>
         sys.error(
-          s"There are ${occurrences.size} groups named $group but we require each group class name to be unique! " +
-            s"Make these group class namers unique to make this assertion run through"
+          s"There are ${occurrences.size} groups named $group but we require each group class name to be unique! "
         )
     }
-
-    (errorCodes.map(convertToDocItem).sortBy(_.code), groups.map(convertToDocItem))
+    errorGroups.map { errorGroup =>
+      ErrorGroupDocItem(
+        errorClass = errorGroup.errorClass,
+        className = errorGroup.fullClassName,
+        explanation = parseErrorGroupAnnotations(errorGroup).explanation,
+      )
+    }
   }
 
-  private def getInstances[T: ru.TypeTag]: Seq[T] =
-    new Reflections(prefixes)
-      .getSubTypesOf(runtimeMirror.runtimeClass(typeOf[T]))
-      .asScala
-      .view
-      .collect {
-        case clazz if !Modifier.isAbstract(clazz.getModifiers) =>
-          clazz.getDeclaredField("MODULE$").get(clazz).asInstanceOf[T]
+  def getErrorCategoryItem(errorCategory: ErrorCategory): ErrorCategoryDocItem = {
+    val mirroredType = runtimeMirror.reflect(errorCategory)
+    val annotations: Seq[ru.Annotation] = mirroredType.symbol.annotations
+    val description = new SettableOnce[String]
+    val resolution = new SettableOnce[String]
+    val retryStrategy = new SettableOnce[String]
+    annotations.foreach { annotation =>
+      lazy val parsedValue = parseAnnotationValue(annotation.tree)
+      getAnnotationTypeName(annotation) match {
+        case DescriptionTypeName =>
+          description.set(parsedValue, DescriptionTypeName)
+        case ResolutionTypeName =>
+          resolution.set(parsedValue, ResolutionTypeName)
+        case RetryStrategyTypeName =>
+          retryStrategy.set(parsedValue, RetryStrategyTypeName)
+        case otherAnnotationTypeName =>
+          throw new IllegalStateException(
+            s"Unexpected annotation of type: $otherAnnotationTypeName"
+          )
       }
-      .toSeq
-
-  private def convertToDocItem(error: ErrorCode): ErrorDocItem = {
-    val ErrorDocumentationAnnotations(deprecation, explanation, resolution) =
-      getErrorDocumentationAnnotations(error)
-
-    ErrorDocItem(
-      className = error.getClass.getName,
-      category = simpleClassName(error.category),
-      hierarchicalGrouping = error.parent,
-      conveyance = error.errorConveyanceDocString,
-      code = error.id,
-      deprecation = deprecation,
-      explanation = explanation,
-      resolution = resolution,
+    }
+    ErrorCategoryDocItem(
+      description = description.get,
+      resolution = resolution.get,
+      retryStrategy = retryStrategy.get,
     )
```

Comment:
Are you sure you don't want to return an `Option` here? Looks a bit unsafe.

Reply:
If there's no documentation we will get `ErrorCategoryDocItem(None, None, None)` which is ok.
Is that what you had in mind?

Btw, I'd like to keep further refactorings to a minimum, as this PR should've been only about using `scala.deprecated` and I already refactored a lot

Reply:
I'm not sure of what `description`'s type is. I might have mistaken it for an `Option`. If you call `get` on `None` you throw an exception. What's the type of `description`, `resolution` and `retryStrategy`?

Reply:
Mh. I've seen the definition of `SettableOnce` below. The statefulness doesn't convince me but I don't have the context to provide clear guidance there, so I'll just resolve this item.